components/raftstore/src/store/peer.rs

```diff
@@ -1513,59 +1529,47 @@ where
-                                // It's safe to read local within its current lease, however, it's not
-                                // safe to renew its lease.
-                                self.last_committed_split_idx = entry.index;
-                                split_to_be_updated = false;
-                            } else if merge_to_be_updated
-                                && ctx.contains(ProposalContext::PREPARE_MERGE)
-                            {
-                                // We committed prepare merge, to prevent unsafe read index,
-                                // we must record its index.
-                                self.last_committed_prepare_merge_idx = entry.get_index();
-                                // After prepare_merge is committed, the leader can not know
-                                // when the target region merges majority of this region, also
-                                // it can not know when the target region writes new values.
-                                // To prevent unsafe local read, we suspect its leader lease.
-                                self.leader_lease.suspect(monotonic_raw_now());
-                                merge_to_be_updated = false;
-                            }
+                    // The admin cmds in `CmdEpochChecker` are proposed by the current leader so we can
+                    // use it to get the split/prepare-merge cmds which was committed just now.
+
+                    // BatchSplit and Split cmd are mutually exclusive because they both change epoch's
+                    // version so only one of them can be proposed and the other one will be rejected
+                    // by `CmdEpochChecker`.
+                    let last_split_idx = self
+                        .cmd_epoch_checker
+                        .last_cmd_index(AdminCmdType::BatchSplit)
+                        .or_else(|| self.cmd_epoch_checker.last_cmd_index(AdminCmdType::Split));
+                    if let Some(idx) = last_split_idx {
+                        if idx <= hs.get_commit() {
+                            // We don't need to suspect its lease because peers of new region that
+                            // in other store do not start election before theirs election timeout
+                            // which is longer than the max leader lease.
+                            // It's safe to read local within its current lease, however, it's not
+                            // safe to renew its lease.
+                            self.last_committed_split_idx = idx;
+                        }
+                    }
+                    let last_prepare_merge_idx = self
+                        .cmd_epoch_checker
+                        .last_cmd_index(AdminCmdType::PrepareMerge);
+                    if let Some(idx) = last_prepare_merge_idx {
+                        if idx <= hs.get_commit() {
```

Comment:
So if the log is not applied, the lease will keep being suspected?

Reply:
Yes. If the `RollbackMerge` is committed after that, the leader can get lease again.

Reply:
But the original implementation only suspect it once.

Reply:
The lease can not renew again because `is_merging` is true.